src/main/java/ru/ralsei/whatcanyousee/logic/MazeGame.java

```diff
@@ -0,0 +1,352 @@
+                        public void run() {
+                            map.draw();
+                            checkIfGameOver();
+                        }
+                    });
+            }
+        }, 0, DELAY, TimeUnit.MILLISECONDS);
+    }
+
+    public void setupListeners() {
+        gameActivity.findViewById(R.id.upButton).setOnClickListener(onClickListener);
+        gameActivity.findViewById(R.id.leftButton).setOnClickListener(onClickListener);
+        gameActivity.findViewById(R.id.rightButton).setOnClickListener(onClickListener);
+        gameActivity.findViewById(R.id.downButton).setOnClickListener(onClickListener);
+        gameActivity.findViewById(R.id.useButton).setOnClickListener(onClickListener);
+        gameActivity.findViewById(R.id.button_show_map).setOnClickListener(onClickListener);
+        gameActivity.findViewById(R.id.button_giveUp_maze).setOnClickListener(onClickListener);
+    }
+
+    /**
+     * On click listener for the maze game.
+     */
+    private View.OnClickListener onClickListener = new View.OnClickListener() {
+        @Override
+        public void onClick(View view) {
+            switch (view.getId()) {
+                case R.id.upButton:
+                    react(MazeGame.Command.UP);
+                    break;
+                case R.id.downButton:
+                    react(MazeGame.Command.DOWN);
+                    break;
+                case R.id.leftButton:
+                    react(MazeGame.Command.LEFT);
+                    break;
+                case R.id.rightButton:
+                    react(MazeGame.Command.RIGHT);
+                    break;
+                case R.id.useButton:
+                    react(MazeGame.Command.USE);
+                    break;
+                case R.id.button_show_map:
+                    ImageView mazeMapImage = gameActivity.findViewById(R.id.image_maze_map);
+
+                    if (mazeMapImage.getVisibility() == View.VISIBLE) {
+                        mazeMapImage.setVisibility(View.INVISIBLE);
+                    } else {
+                        mazeMapImage.setVisibility(View.VISIBLE);
+                    }
+                    break;
+                case R.id.button_giveUp_maze: {
+                    gameActivity.runOnUiThread(new Runnable() {
+                        @Override
+                        public void run() {
+                            gameActivity.getGameplayHandler().onMazeGameLost("You gave up.");
+                        }
+                    });
+                    break;
+                }
+                default:
+                    break;
+            }
+        }
+    };
+
+    /**
+     * User commands from the screen buttons.
+     */
+    public enum Command {
+        DOWN, LEFT, UP, RIGHT, USE
+    }
+
+    /**
+     * Related coordinates of all 8 cells around the player.
+     */
+    private MazeGameMap.Coordinates[] vec8 = {
+            new MazeGameMap.Coordinates(-1, 0),
```

Comment:
Why is this one an array, and the one below a list?

Reply:
Because vec8 is fixed, and vec4 supposed to be used in random order (therefore shuffling, therefore list), but yes, that's not obvious from the code.